### PR TITLE
Migrate from PCL to .NET Standard

### PR DESCRIPTION
**Pull request by MatkovIvan**
.NET Standard can be thought of as the next generation of PCL. Look like it's a better way for cross-platform code part. According [this](https://docs.microsoft.com/en-us/dotnet/articles/standard/library) PCL Profile259 compatible with the .NET Standard 1.0.
This change doesn't break capability with PCL applications.

This PR most like topic to discuss than ASAP to merge. Therefore, feel free to add remarks/suggestion.

### CODE REVIEW

File: nuget/MacMobileCenter.nuspec [hunk 1]
```diff
@@ -11,18 +11,23 @@
     <projectUrl>https://aka.ms/telgml</projectUrl>
     <licenseUrl>https://aka.ms/vbgfx2</licenseUrl>
     <iconUrl>https://aka.ms/p2wifj</iconUrl>
+    <dependencies>
+        <group targetFramework="netstandard1.0">
+            <dependency id="NETStandard.Library" version="1.6.0" />
```
Comment (achocron) [inline nuget/MacMobileCenter.nuspec]: Why depend on version 1.6.0 instead of 1.0.0? Won't that worsen compatibility?

**Comment by achocron** [comment]
When I tried to make a Xamarin.Forms application that targets 259 and import a nuget package that has a dependency on .netstandard1.0, I got this error:
`Could not install package 'Microsoft.NETCore.Runtime.Native 1.0.2-rc2-24027'. You are trying to install this package into a project that targets '.NETPortable,Version=v4.5,Profile=Profile259', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.`

So it looks like it would break compatibility.

File: nuget/MacMobileCenter.nuspec (hunk 1 above)
Comment (MatkovIvan) [inline nuget/MacMobileCenter.nuspec]: It is compatible with .NET Standard 1.0. See https://www.nuget.org/packages/NETStandard.Library/

**Comment by MatkovIvan** [comment]
@achocron Mobile Center doesn't have `Microsoft.NETCore.Runtime.Native` dependency. I rechecked this case now and it works. Maybe you tried link to it in your test project?

**Comment by guperrot** [comment]
CI Error is a test failing:

```
2017-05-19T23:47:03.8821028Z Failed   ChannelInvokesFailedToSendLogEventAfterDisabling
2017-05-19T23:47:03.8821028Z ##[error]Error Message:
2017-05-19T23:47:03.8821028Z 
2017-05-19T23:47:03.8821028Z ##[error]   Test method Microsoft.Azure.Mobile.Test.Channel.ChannelTest.ChannelInvokesFailedToSendLogEventAfterDisabling threw exception: 
2017-05-19T23:47:03.8821028Z ##[error]System.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.
2017-05-19T23:47:03.8821028Z ##[error]Stack Trace:
2017-05-19T23:47:03.8821028Z ##[error]    at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)
2017-05-19T23:47:03.8821028Z ##[error]   at System.Threading.SemaphoreSlim.Release()
2017-05-19T23:47:03.8821028Z ##[error]   at Microsoft.Azure.Mobile.Utils.Synchronization.StatefulMutex.Unlock() in C:\Users\mobilecenter\agent\_work\2\s\SDK\********\Microsoft.Azure.Mobile.Windows.Shared\Utils\Synchronization\StatefulMutex.cs:line 82
2017-05-19T23:47:03.8821028Z ##[error]   at Microsoft.Azure.Mobile.Channel.Channel.Enqueue(Log log) in C:\Users\mobilecenter\agent\_work\2\s\SDK\********\Microsoft.Azure.Mobile.Windows.Shared\Channel\Channel.cs:line 137
2017-05-19T23:47:03.8821028Z ##[error]   at Microsoft.Azure.Mobile.Test.Channel.ChannelTest.ChannelInvokesFailedToSendLogEventAfterDisabling() in C:\Users\mobilecenter\agent\_work\2\s\Tests\Microsoft.Azure.Mobile.Test.Windows\Channel\ChannelTest.cs:line 194
```

**Comment by guperrot** [comment]
@achocron @MatkovIvan I ran publish internal with this PR to produce the nuget packages into our internal repos.

You can start testing against version `0.12.1-r0002-1c917ba` of our nugets.

**Comment by MatkovIvan** [comment]
This test fail should be fixed in https://github.com/Microsoft/mobile-center-sdk-dotnet/pull/317. Merged to verify that. Locally tests run succeeded.

**Comment by Oleksandr-Tokmakov** [comment]
Guys, when can I expect for new release with .netstandart support? I really need it asap.

File: build.cake [hunk 2]
```diff
@@ -337,25 +337,38 @@ Task("UITest").IsDependentOn("RestoreTestPackages").Does(() =>
 
 // Pack NuGets for appropriate platform
 Task("NuGet")
```
Comment (achocron) [inline build.cake]: Why these changes?

File: nuget/WindowsMobileCenterCrashes.nuspec [hunk 3]
```diff
@@ -19,23 +19,23 @@
...
-    <file src="SDK/MobileCenterCrashes/Microsoft.Azure.Mobile.Crashes.UWP/bin/Reference/Microsoft.Azure.Mobile.Crashes.dll" target="ref/uap10.0" />
+    <file src="$windows_dir$/Microsoft.Azure.Mobile.Crashes.targets" target="build/uap10.0" />
+    <file src="$windows_dir$/Microsoft.Azure.Mobile.Crashes.dll" target="ref/uap10.0" />
     
     <!--x86 -->
-    <file src="SDK/MobileCenterCrashes/Microsoft.Azure.Mobile.Crashes.UWP/bin/x86/Release/Microsoft.Azure.Mobile.Crashes.dll" target="runtimes/win10-x86/lib/uap10.0" />
-    <file src="Release/WatsonRegistrationUtility/WatsonRegistrationUtility.dll" target="runtimes/win10-x86/native"/>
-    <file src="Release/WatsonRegistrationUtility/WatsonRegistrationUtility.winmd" target="runtimes/win10-x86/lib/uap10.0" />
+    <file src="$windows_dir$/Microsoft.Azure.Mobile.Crashes.dll" target="runtimes/win10-x86/lib/uap10.0" />
+    <file src="$windows_dir$/WatsonRegistrationUtility.dll" target="runtimes/win10-x86/native"/>
```
Comment (achocron) [inline nuget/WindowsMobileCenterCrashes.nuspec]: Since the $windows_dir$ doesn't actually include the "x86" "ARM" or "x64" folder, you still need to include it. (See the combined nuspecs for reference)

**Comment by achocron** [comment]
@Oleksandr-Tokmakov We are working on it now - I'll let you know once it's released

File: nuget/WindowsMobileCenterCrashes.nuspec (hunk 3 above)
Comment (MatkovIvan) [inline nuget/WindowsMobileCenterCrashes.nuspec]: Yeah, I missed it.

File: build.cake (hunk 2 above)
Comment (MatkovIvan) [inline build.cake]: To replacing `$windows_dir$` and others in nuspecs

**Comment by Oleksandr-Tokmakov** [comment]
@achocron  ok, thanks! By the way I downloaded sources from GitHub, changed framework to .netstandard 1.6 and added compiled dlls to my project instead of nugget packages. Everything works, took 10 minutes.

**Comment by achocron** [comment]
@Oleksandr-Tokmakov Just a heads up that we have released Mobile Center 0.13.0 on NuGet, which supports .NET Standard! https://github.com/Microsoft/mobile-center-sdk-dotnet/releases/tag/0.13.0
